File: src/types/events.ts
// File: /types/events.ts
import { BN } from "@coral-xyz/anchor";
import { PublicKey } from "@solana/web3.js";

export type WebhookEventType =
  | "campaign_created"
  | "donation_received"
  | "user_initialized"
  | "campaign_updated"
  | "funds_withdrawn";

export const eventTypeMap = {
  CREATE_CAMPAIGN: "campaign_created",
  DONATE: "donation_received",
  INITIALIZE: "user_initialized",
  UPDATE_CAMPAIGN: "campaign_updated",
  WITHDRAW_FUNDS: "funds_withdrawn",
} as const;

export type RawEventType = keyof typeof eventTypeMap;

export interface RawEvent {
  parsedData: {
    type: RawEventType;
    data: Record<string, unknown>;
  };
  blockTime?: number;
  accounts?: Record<string, unknown>;
}
export const VALID_EVENT_TYPES = Object.values(eventTypeMap);

export interface WebhookEventMetadata {
  title?: string;
  category?: string;
  targetAmount?: string;
  organization?: string;
  startDate?: number;
  endDate?: number;
  isUrgent?: boolean;
  amount?: string;
  paymentMethod?: string;
  name?: string;
  updates?: {
    description?: string | null;
    imageUrl?: string | null;
    endDate?: number | null;
    status?: string | null;
    isUrgent?: boolean | null;
  };
}

export interface WebhookEventPayload {
  metadata: WebhookEventMetadata;
  blockTime: number;
  accounts: Record<string, unknown>;
  parsedData: {
    type: string;
    data: Record<string, unknown>;
  };
}
export interface WebhookEvent {
  type: "webhook";
  eventType: WebhookEventType;
  timestamp: string;
  payload: WebhookEventPayload;
}

export interface SystemEvent {
  type: "connection" | "ping";
  timestamp: string;
  status?: "connected";
  clientCount?: number;
}

export type ServerEvent = WebhookEvent | SystemEvent;

// API Response Types
export interface WebhookResponse {
  received: boolean;
  eventsProcessed: number;
  totalEvents: number;
  skippedEvents: number;
  timestamp: string;
}

export interface WebhookError {
  message: string;
  error: string;
}

// Validation Types for specific events
export interface CampaignCreatedEvent {
  accounts: {
    campaign: PublicKey;
    authority: PublicKey;
  };
  data: {
    title: string;
    category: string;
    targetAmount: BN;
    organizationName: string;
    startDate: BN;
    endDate: BN;
  };
}

export interface DonationReceivedEvent {
  accounts: {
    donation: PublicKey;
    campaign: PublicKey;
    donor: PublicKey;
  };
  data: {
    amount: BN;
    timestamp: BN;
    paymentMethod: string;
  };
}

export interface UserInitializedEvent {
  accounts: {
    user: PublicKey;
    authority: PublicKey;
  };
  data: {
    name: string;
  };
}

export interface EventStreamState {
  isConnected: boolean;
  events: WebhookEvent[];
  lastEventTimestamp: string | null;
  error: Error | null;
}

File: anchor/programs/ripple/src/events.rs
// File: src/events.rs
use anchor_lang::prelude::*;
use crate::state::*;

// Re-export events from lib.rs
pub use crate::{ CampaignCreated, DonationReceived, CampaignUpdated, FundsWithdrawn, BadgeAwarded };

// Helper functions to emit events
pub fn emit_campaign_created(
    campaign: Pubkey,
    authority: Pubkey,
    title: String,
    category: CampaignCategory,
    target_amount: u64
) {
    emit!(CampaignCreated {
        campaign,
        authority,
        title,
        category,
        target_amount,
    });
}

pub fn emit_donation_received(
    donation: Pubkey,
    campaign: Pubkey,
    donor: Pubkey,
    amount: u64,
    payment_method: PaymentMethod
) {
    emit!(DonationReceived {
        donation,
        campaign,
        donor,
        amount,
        payment_method,
    });
}

pub fn emit_campaign_updated(
    campaign: Pubkey,
    authority: Pubkey,
    new_status: Option<CampaignStatus>
) {
    emit!(CampaignUpdated {
        campaign,
        authority,
        new_status,
    });
}

pub fn emit_funds_withdrawn(campaign: Pubkey, recipient: Pubkey, amount: u64) {
    emit!(FundsWithdrawn {
        campaign,
        recipient,
        amount,
    });
}

pub fn emit_badge_awarded(user: Pubkey, badge_type: BadgeType, timestamp: i64) {
    emit!(BadgeAwarded {
        user,
        badge_type,
        timestamp,
    });
}
